gh-103743: Add PyUnstable_Object_GC_NewWithExtraData

jbradaric opened the pull request:
<!-- gh-issue-number: gh-103743 -->
* Issue: gh-103743
<!-- /gh-issue-number -->

jbradaric committed:
gh-103743: Add PyUnstable_Object_GC_NewWithExtraData (#103743)

Eclips4 commented on Doc/c-api/gcsupport.rst:
```diff
@@ -65,6 +65,20 @@ rules:
+   .. warning::
+       The function is marked as unstable because the final mechanism
+       for reserving extra data after an instance is not yet decided.
+       Once `PEP-697<https://peps.python.org/pep-0697/>`_ is
```
```suggestion
       Once :pep:`697` is
```

jbradaric replied:
Thanks, I was wondering how that's done :+1:  Fixed

ambv committed:
Merge branch 'main' into jbradaric/obj-new-with-extra-data

encukou committed:
Don't suggest PEP 697 will be a solution

encukou committed:
Move declaration to Include/cpython

That way it doesn't need the Py_LIMITED_API guard

encukou committed:
Style nitpicks

encukou committed:
Clarify the docs a bit

encukou committed:
Convert the test class to a heap type

encukou committed:
Initialize the memory

This is a further workaround: this time for the fact that
we can't rely on tp_alloc, which zeroes the memory (including
memory for subclasses, like slots in the test) for normal classes.

For the test type, this means the instances aren't initialized twice
(once in PyUnstable_Object_GC_NewWithExtraData, once again with
PyObject_Init after zeroing). Initializing twice throws off the global
refcount total in debug builds.

erlend-aasland commented on Doc/c-api/gcsupport.rst:
```diff
@@ -59,12 +59,31 @@ rules:
+   bytes at the end of the object (at offset
+   :c:member:`~PyTypeObject.tp_basicsize`).
+   The allocated memory is initialized: to zero, except for the
+   :c:type:`Python object header <PyObject>`.
```
I find it weird with the colon. How about this? (Also, SemBr)

```suggestion
   The allocated memory is initialized to zeros,
   except for the :c:type:`Python object header <PyObject>`.
```

erlend-aasland commented on Doc/c-api/gcsupport.rst:
```diff
@@ -59,12 +59,31 @@ rules:
+       for reserving extra data after an instance is not yet decided.
+       For allocating a variable number of fields, prefer using
+       :c:type:`PyVarObject` and :c:member:`~PyTypeObject.tp_itemsize`
+       instead.
```
Nit: 3-space indent for .rst

```suggestion
      The function is marked as unstable because the final mechanism
      for reserving extra data after an instance is not yet decided.
      For allocating a variable number of fields, prefer using
      :c:type:`PyVarObject` and :c:member:`~PyTypeObject.tp_itemsize`
      instead.
```

erlend-aasland commented on Misc/NEWS.d/next/C API/2023-04-24-10-31-59.gh-issue-103743.2xYA1K.rst:
```diff
@@ -0,0 +1,2 @@
+Add ``PyUnstable_Object_GC_NewWithExtraData`` function that can be used to
```
```suggestion
Add :c:func:`PyUnstable_Object_GC_NewWithExtraData` function that can be used to
```

erlend-aasland reviewed (commented):
LGTM. Minor nits regarding docs.

jbradaric committed:
Doc style fixes

Fix indentation and reword some parts to sound more natural.

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>
